app/com/lynxanalytics/biggraph/frontend_operations/ImportOperations.scala

```diff
@@ -151,27 +145,113 @@ class ImportOperations(env: SparkFreeEnvironment) extends OperationRegistry {
+        .load
+        .cache
+
+      def getBy(columns: Seq[String], key: String, id: String = null) = {
+        import spark.implicits._
+        val available = columns.filter(df.columns.contains(_))
+        val kinds: Set[String] = available.map { column =>
+          df
+            .select(get_json_object(col(column), "$['" + key + "']"))
+            .dropDuplicates
+            .collect.map(r => r.getString(0)).toSet
+        }.reduce(_ ++ _)
+        kinds.map { k =>
+          val parsed = spark.read.json(
+            available
+              .map { column =>
+                df.filter(get_json_object(col(column), "$['" + key + "']") === k)
+                  .select(col(column)).as[String]
+              }
+              .reduce(_.union(_)))
+          val dedup = if (id != null) parsed.dropDuplicates(id) else parsed
+          graph_operations.ImportDataFrame.run(dedup)
+        }
+      }
+
+      for (table <- getBy(Seq("src", "dst", "node"), "<labels>", "<id>")) {
+        mergeInto(project, table.toProject.viewer)
+      }
+      assert(
+        project.vertexAttributes.contains("<id>"),
+        s"No 'src', 'dst', or 'node' found in query response: ${df.columns}")
+      for (table <- getBy(Seq("rel"), "<rel.type>")) {
+        addEdges(project, table.toProject.viewer)
+      }
+      val state = json.Json.toJson(project.state).toString
+      ImportResult(state, this.settingsString())
+    }
+
+    def addEdges(project: ProjectEditor, edges: ProjectViewer): Unit = {
+      import com.lynxanalytics.biggraph.graph_util.Scripting._
+      val idAttr = project.vertexAttributes("<id>").asString
+      val srcAttr = edges.vertexAttributes("<source.id>").asString
+      val dstAttr = edges.vertexAttributes("<target.id>").asString
+      val imp = graph_operations.ImportEdgesForExistingVertices.run(
+        idAttr, idAttr, srcAttr, dstAttr)
+
+      if (project.edgeBundle == null) {
+        project.edgeBundle = imp.edges
+        for ((name, attr) <- edges.vertexAttributes) {
+          project.edgeAttributes(name.replace(".", "_")) = attr.pullVia(imp.embedding)
+        }
+        return
+      }
+      val idUnion = {
+        val op = graph_operations.VertexSetUnion(2)
+        op(op.vss, Seq(project.edgeBundle.idSet, imp.edges.idSet)).result
+      }
+      val ebUnion = {
+        val op = graph_operations.EdgeBundleUnion(2)
+        op(
+          op.ebs, Seq(project.edgeBundle, imp.edges.entity))(
+            op.injections, idUnion.injections.map(_.entity)).result.union
+      }
+      val oldAttrs = project.edgeAttributes.toMap
+      project.edgeBundle = ebUnion
+      project.edgeAttributes = unifyAttributes(
+        oldAttrs.map {
+          case (name, attr) => name -> attr.pullVia(idUnion.injections(0).reverse)
+        },
+        edges.vertexAttributes.map {
+          case (name, attr) => name -> attr.pullVia(idUnion.injections(1).reverse.concat(imp.embedding))
+        })
```

Comment:
Can't we do the unions on the df level and convert to edges only once? I'm worried this is very slow. Also pretty complex.

Reply:
> Can't we do the unions on the df level and convert to edges only once? I'm worried this is very slow. Also pretty complex.

Both with edges and nodes the issue is that different types (labels) can have different sets of attributes. So `spark.read.json()` fails on the whole dataset. To make it work I split by label first, run `spark.read.json()` on each type and then union them back, unifying the attributes here.

But now that I think about it I'm not sure how the JSON -> DF parser deals with sparsely defined fields. Is that any different from the situation with different types of nodes? If it just guesses the schema based on the first 10 rows, that is not great. Maybe this can be controlled? I'll take another look.

Reply:
> But now that I think about it I'm not sure how the JSON -> DF parser deals with sparsely defined fields. Is that any different from the situation with different types of nodes? If it just guesses the schema based on the first 10 rows, that is not great. Maybe this can be controlled? I'll take another look.

It reads the whole input to determine the schema. (https://github.com/apache/spark/blob/v2.4.7/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala) It works fine with the whole heterogenous node set from Neo4j.